### PR TITLE
feat: add net http instrumentation hooks config

### PR DESCRIPTION
**Pull request by nozik**
Add support for `request_hook` and `response_hook` that allow adding user-customizable attributes to the span. This capability is common across http client instrumentations in other languages (see https://github.com/open-telemetry/opentelemetry-js/tree/main/experimental/packages/opentelemetry-instrumentation-http and https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation/opentelemetry-instrumentation-urllib3 in node and python, respectively).

For example, this capability can be used to collect data from request/response headers.

### CODE REVIEW

**Comment by plantfansam** [comment]
Hi @nozik, and thanks for your contribution! Can you write a little bit more about what this would be used for and link to any relevant sections of the OpenTelemetry spec (there may be none, just want to check!).

**Comment by nozik** [comment]
@plantfansam Done :)

File: instrumentation/net_http/README.md
Comment (nozik) [reply instrumentation/net_http/README.md:41]: Yes, I've shared a couple of examples above:

- Node - [link](https://github.com/open-telemetry/opentelemetry-js/blob/b8915092bc15e0b38b0388463e05308a438282ac/experimental/packages/opentelemetry-instrumentation-http/src/types.ts#L111)
- Python - [link](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation/opentelemetry-instrumentation-urllib3#requestresponse-hooks)

File: instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb
Comment (nozik) [reply instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb:49]: A wrong number of arguments will throw an exception - which will be caught and reported in a debug log. I can use `arity` during the option validation but that seems like an overkill, IMHO.

**Comment by nozik** [comment]
1. Cardinality issues - I generally agree, but collecting additional data has so many use cases, and it's really up to the user to do that with care.
2. We intend to make similar contribution for other instrumentations (not only http client) :)

**Review by ericmustin — APPROVED** [review]
lgtm, looks like a rubocop complaint but beside that i think this is fine. Still getting my feet wet wrt this repo (ive been on pto for awhile), so wouldnt mind if @plantfansam or @arielvalentin could review as well.

File: instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb
Comment (ericmustin) [reply instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb:49]: fair enough, i agree, probably fine for now

File: instrumentation/net_http/README.md
Comment (nozik) [reply instrumentation/net_http/README.md]: Good catch :)

**Comment by nozik** [comment]
@ericmustin Thank you!

Still getting the `Metrics/AbcSize` on the `def request(req, body = nil, &block)`, and I'm not sure if it makes sense. Any suggestion on how to resolve this? I did not add any assignment, branch or condition to the function, just a method call. Thanks!

File: instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb
Comment (plantfansam) [inline instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb:34]: What do you think about keeping the `propagation.inject` call up here, then, on the next line, invoking `safe_execute_hook if instrumentation_config[:request_hook]`. I'm not sure why we'd group together `inject_context_and_extract_data` into a single method (but maybe I'm missing something!).

File: instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb
Comment (plantfansam) [inline instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb:88]: Same deal here — maybe `safe_execute_hook if instrumentation_config[:response_hook]`?

**Review by plantfansam — COMMENTED** [review]
Thanks so much for the contribution — seems like a great feature to me! Had a question about style, which I make inline.

File: instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb
Comment (arielvalentin) [inline instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb:58]: Did you consider using [`OpenTelemetry.handle_error`](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/api/lib/opentelemetry.rb#L44)?

The advantage of using the `handle_error` is that it provides access to the `Error` and make it possible to inspect details like the backtrace.

File: instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb
Comment (arielvalentin) [reply instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb:34]: +1

File: instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb
Comment (arielvalentin) [inline instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb:149]: Unless I am missing something, are there any test cases that cover the exception flow for when the request and response hooks fail?

**Review by arielvalentin — COMMENTED** [review]
Thank you for taking the time to bring in these ideas from other Language SDKs. 

I have some requests for changes e.g. adding tests for error handling.

File: instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb
Comment (nozik) [reply instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb:34]: Hi @plantfansam @arielvalentin  , the reason I did it was to avoid `Metrics/AbcSize` lint error. For some reason only in Ruby 2.7 it would fail. I'd appreciate any other suggestion you may to workaround this error. Thanks!

File: instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb
Comment (nozik) [reply instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb:88]: I'm checking the condition inside the function, so I preferred to not to do it twice and keep the method safer. I don't mind changing it though, if you prefer.

File: instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb
Comment (nozik) [reply instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb:58]: Done

File: instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb
Comment (nozik) [reply instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb:149]: Added

File: instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb
Comment (arielvalentin) [reply instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb:34]: @plantfansam we should probably take a look at the ABCSize rule and see if we should relax it globally for instrumentations that use prepended modules.

File: instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb
Comment (arielvalentin) [inline instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb:189]:
```suggestion
            request_hook: lambda { |_span, _request, _request_body| raise StandardError, 'err1' },
```

File: instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb
Comment (arielvalentin) [inline instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb:190]:
```suggestion
            response_hook: lambda { |_span, _response| raise StandardError, 'err2' },
```

File: instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb
Comment (arielvalentin) [inline instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb:170]:
```suggestion
            request_hook: lambda { |_span| nil },
```

File: instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb
Comment (arielvalentin) [inline instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb:171]:
```suggestion
            response_hook: lambda { |_span| nil },
```

**Review by arielvalentin — COMMENTED** [review]
I have some suggestions for you. 

- I'd like to avoid printing to stdout in test cases
- one of the test cases covering for errors returned a StandardError instead of raising it
- Please consider adding an assertion that the error handler was called.

File: instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb
Comment (nozik) [reply instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb]: Done

File: instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb
Comment (nozik) [reply instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb]: Done

File: instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb
Comment (nozik) [reply instrumentation/net_http/test/opentelemetry/instrumentation/net/http/instrumentation_test.rb]: Done

**Comment by nozik** [comment]
@arielvalentin I've applied your suggestions, except for making sure the error handler is called. Is there a standard way that you may suggest for me to do this? I'm new to this repo after all. Thanks!

**Comment by arielvalentin** [comment]
> I've applied your suggestions, except for making sure the error handler is called. Is there a standard way that you may suggest for me to do this? I'm new to this repo after all. Thanks!

It looks like we do not use the error handler much in instrumentations. No wonder there aren't any examples!

Here are some examples of how its used in tests in the SDK repo:

https://github.com/open-telemetry/opentelemetry-ruby/blob/81707ad23d4d0aba72d41fb772f4edc913b2ffd2/sdk/test/opentelemetry/sdk_test.rb#L16

**Comment by nozik** [comment]
@arielvalentin Done, thanks for the guidance.

Are we good to go to merge? Once we are, I'll prepare similar PRs for the rest of the HTTP clients based on this one. Thanks

**Review by arielvalentin — APPROVED** [review]
Congratulations on your first contribution! 

Thank you again for being patient and incorporating my feedback! I am excited to see that you are interested in following up with more PRs and look forward to future contributions.  

With respect to merging, an implicit practice before merging is that resolving open conversations with other maintainers. It looks like the only open threads you have are with @plantfansam. 

Once he gives a ✅ we will merge it.

File: instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb
Comment (arielvalentin) [reply instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb:34]: Related https://github.com/open-telemetry/opentelemetry-ruby-contrib/issues/66

Comment (plantfansam) [reply instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb:34]: @nozik you can add `# rubocop:disable Metrics/AbcSize` as a comment at the end of the line that Rubocop complains about, and that will fix it!

File: instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb
Comment (plantfansam) [reply instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb:88]: I would rather check the condition prior to entering the method 😄

File: instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb
Comment (nozik) [reply instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb]: @plantfansam Done!

File: instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb
Comment (nozik) [reply instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb]: Done

**Comment by nozik** [comment]
@plantfansam I've applied the requested changes. Thanks!

**Review by plantfansam — APPROVED** [review]
LGTM! Thanks so much for your contribution!

**Pull request merged**
